docs/content/sips/008-using-oci-registries.md

```diff
@@ -0,0 +1,298 @@
+entities in OCI). Specifically, the media type used in this implementation
+for a Spin application is `application/vnd.fermyon.spin.application.v1+config`,
+and each Wasm module and static asset from the Spin components becomes an individual
+*layer* in the resulting registry entity. Because each file and Wasm module becomes
+a separate *layer*, we can efficiently de-duplicate and distribute applications.
+
+The remaining question is representing the Spin application definition — Spin
+introduced the internal representation of a *locked application* - an
+intermediate representation of a Spin application that *can* have a way to
+content-address Wasm modules and static assets. The implementation for this SIP
+uses the Spin locked application as the OCI configuration object.
+
+Below is an example of a Spin application with one component and one static
+asset, its OCI manifest, locked application, and resulting local directory
+structure when pulling the application locally.
+
+Consider the following `spin.toml` file for the application:
+
+```toml
+spin_version = "1"
+authors = ["Radu Matei <radu.matei@fermyon.com>"]
+description = ""
+name = "github-stars-webhook"
+trigger = { type = "http", base = "/" }
+version = "0.1.0"
+
+[[component]]
+id = "github-star-webhook"
+source = "target/spin-http-js.wasm"
+files = ["my-file.json"]
+allowed_http_hosts = ["https://hooks.slack.com"]
+[component.trigger]
+route = "/..."
+```
+
+Distributing it to the GitHub Container Registry:
+
+```bash
+$ spin oci push ghcr.io/radu-matei/spin-example:v1
+INFO spin_publish::oci::client: Pushed "https://ghcr.io/v2/radu-matei/spin-example/manifests/sha256:8f86a27fbc457416701c4d18680083f598076d0a52dca2a5936e92754a845ed1"
+
+$ spin oci pull ghcr.io/radu-matei/spin-example:v1
+INFO spin_publish::oci::client: Pulled ghcr.io/radu-matei/spin-example:v1@sha256:8f86a27fbc457416701c4d18680083f598076d0a52dca2a5936e92754a845ed1
+```
+
+This operation created the following local cache directory structure:
+
+```bash
+$ tree /Users/radu/Library/Application\ Support/fermyon/registry
+└── oci
+    ├── data
+    │   └── sha256:a4699e4f9ef3f4922f38f0d017aa26438908f38caf020a739e0ee27fe796eb02
+    ├── manifests
+    │   └── ghcr.io
+    │       └── radu-matei
+    │           └── spin-example
+    │               └── v1
+    │                   ├── config.json
+    │                   └── manifest.json
+    └── wasm
+        └── sha256:55c29ad4b0ad0c6bd8ec1ffc8f04e63342e5901280037ef706b1b114475d3cbb
+```
+
+Looking at `manifest.json`, we see the top-level media type for the artifact
+configuration to be `application/vnd.fermyon.spin.application.v1+config`
+
+```json
+{
+  "schemaVersion": 2,
+  "config": {
+    "mediaType": "application/vnd.fermyon.spin.application.v1+config",
+    "digest": "sha256:b36160facea3076ad136c09bd4975a429805945ad313b4674363841d5a7f66a0",
+    "size": 643
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+wasm",
```

Comment:
We might consider using the 
`application/vnd.module.wasm.content.layer.v1+wasm` content media type for Wasm modules; this might make more sense as we could add `wasm.component` in the future.

Reply:
is there any standardization on wasm oci layer names  yet (from Docker for example)?

Reply:
There are a few implementations out there that distribute Wasm _modules_ using OCI:

- [`wasm-to-oci`](https://github.com/engineerd/wasm-to-oci) and [`oci-distribution`](https://github.com/krustlet/oci-distribution) use `application/vnd.wasm.content.layer.v1+wasm`
- [`solo-io/wasm/spec`](https://github.com/solo-io/wasm/blob/master/spec/spec.md) uses `application/vnd.module.wasm.content.layer.v1+wasm`
- Docker's preview appears to distribute them as `application/vnd.docker.container.image.v1+json`

In short, there is no clear standardization at this point.

Reply:
Added a note in the SIP referencing the above.

Reply:
Thanks for laying that out. I wonder if there should be cross organization work to look at standardizing this